File: marketdna-backend/app/models/validation.py
"""Pydantic models for the Pattern DNA Validation Report."""
from pydantic import BaseModel


class OOSSplitResult(BaseModel):
    pattern: str
    n_symbols: int              # stocks with ≥5 years of history
    is_n: int                   # in-sample qualified detections
    oos_n: int                  # out-of-sample qualified detections
    is_success_rate: float      # % of IS detections that succeeded
    oos_success_rate: float     # % of OOS detections that succeeded
    is_avg_ret_21d: float       # avg 21-day forward return in IS period
    oos_avg_ret_21d: float      # avg 21-day forward return in OOS period
    degradation_pct: float      # (IS_SR - OOS_SR) / IS_SR × 100
    passes: bool                # OOS degradation < 20%


class DecileRow(BaseModel):
    rank: str                   # "Q1 (Best DNA)" … "Q5 (Worst DNA)"
    n_symbols: int
    avg_dna_score: float        # mean IS DNA score in this group
    oos_avg_ret_21d: float      # mean OOS 21-day return when pattern detected
    n_detections: int           # OOS detections in this group


class DecileResult(BaseModel):
    pattern: str
    rows: list[DecileRow]
    spread_pct: float           # Q1 return − Q5 return
    monotonic: bool             # does return improve from Q5 → Q1?
    passes: bool                # spread > 2% and monotonic


class ConfBin(BaseModel):
    label: str                  # "No volume (0)", "Partial (1)", "Full (2+)"
    n: int
    success_rate: float
    avg_ret_21d: float


class CalibrationResult(BaseModel):
    pattern: str
    bins: list[ConfBin]
    monotonic: bool             # does success rate rise with volume confirmation?
    passes: bool                # monotonic = True and spread between full/none > 5%


class ValidationSummary(BaseModel):
    pattern: str
    step2_oos: bool | None
    step3_decile: bool | None
    step4_calibration: bool | None
    verdict: str                # "VALIDATED" | "PARTIAL" | "UNVALIDATED" | "INSUFFICIENT_DATA"


class ValidationReport(BaseModel):
    generated_at: str
    is_bars: int                # 756 (~3 years)
    oos_bars: int               # 504 (~2 years)
    min_occurrences: int        # 12
    n_symbols_total: int
    n_symbols_sufficient: int   # stocks with ≥5 years history

    oos_results: list[OOSSplitResult]
    decile_results: list[DecileResult]
    calibration_results: list[CalibrationResult]
    summary: list[ValidationSummary]

File: marketdna-web/src/types/validation.ts
export interface OOSSplitResult {
  pattern: string
  n_symbols: number
  is_n: number
  oos_n: number
  is_success_rate: number
  oos_success_rate: number
  is_avg_ret_21d: number
  oos_avg_ret_21d: number
  degradation_pct: number
  passes: boolean
}

export interface DecileRow {
  rank: string
  n_symbols: number
  avg_dna_score: number
  oos_avg_ret_21d: number
  n_detections: number
}

export interface DecileResult {
  pattern: string
  rows: DecileRow[]
  spread_pct: number
  monotonic: boolean
  passes: boolean
}

export interface ConfBin {
  label: string
  n: number
  success_rate: number
  avg_ret_21d: number
}

export interface CalibrationResult {
  pattern: string
  bins: ConfBin[]
  monotonic: boolean
  passes: boolean
}

export interface ValidationSummary {
  pattern: string
  step2_oos: boolean | null
  step3_decile: boolean | null
  step4_calibration: boolean | null
  verdict: 'VALIDATED' | 'PARTIAL' | 'UNVALIDATED' | 'INSUFFICIENT_DATA'
}

export interface ValidationReport {
  generated_at: string
  is_bars: number
  oos_bars: number
  min_occurrences: number
  n_symbols_total: number
  n_symbols_sufficient: number
  oos_results: OOSSplitResult[]
  decile_results: DecileResult[]
  calibration_results: CalibrationResult[]
  summary: ValidationSummary[]
}
